### ISSUE TITLE
How can I set the metadata of each blob using the uploadasync method

### ISSUE THREAD

**Issue by sunilprabha** [Azure/azure-sdk-for-net#17947]
I am using this class to upload data to  Azure and when uploading the block blob i want to set the metadata of each blob. 
Cannot find a method to set the metadata of the blob.
Regards
Sunil 
[Enter feedback here]

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 025a85b9-4978-0540-b42e-5d2f516001f4
* Version Independent ID: ba11ea96-5e96-2212-d8d0-58178380cb22
* Content: [TransferManager.UploadAsync Method (Microsoft.Azure.Storage.DataMovement) - Azure for .NET Developers](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.storage.datamovement.transfermanager.uploadasync?view=azure-dotnet)
* Content Source: [xml/Microsoft.Azure.Storage.DataMovement/TransferManager.xml](https://github.com/Azure/azure-docs-sdk-dotnet/blob/master/xml/Microsoft.Azure.Storage.DataMovement/TransferManager.xml)
* Service: **storage**
* GitHub Login: @CamSoper
* Microsoft Alias: **casoper**

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by amnguye** [comment]
Hello! You can set the metadata when uploading a blob by passing the metadata through the `BlobUploadOptions`

Here's the API
```
public virtual async Task<Response<BlobContentInfo>> UploadAsync(
            Stream content,
            BlobUploadOptions options,
            CancellationToken cancellationToken = default)
 ```
 
Here's a tiny snippet of how to do so
```
BlobUploadOptions options = new BlobUploadOptions() { Metadata = blobMetadata };
await UploadAsync(streamContent, options);
```

**Comment by sunilprabha** [comment]
Hello Amadna,
I see your snippet , however don't see in my object metadata property .  Any pointer is appreciated.

[cid:05b286d4-16c6-4fc1-88a4-27386b5056bf]
Regards
Sunil Prabha
404 542 5536
________________________________

**Comment by amnguye** [comment]
Hi could you provide me the exact code snippet you are doing in order to upload the metadata. Also how are you checking for the metadata? Are you calling `GetProperties` ?

**Comment by sunilprabha** [comment]
Hi,
Here is the method, it is posting the files, however, the problem is i don't see a way to set the meta data without getting the reference to the blob .What am I missing.


using Microsoft.Azure.Storage.DataMovement;
using Microsoft.Azure.Storage;
using Microsoft.Azure.Storage.Blob;

public static async Task TransferLocalFileToAzureBlob(Microsoft.Azure.Storage.CloudStorageAccount account)
        {
            string localFilePath = GetSourcePath();
            CloudBlockBlob blob = GetBlob(account);
            TransferCheckpoint checkpoint = null;
            SingleTransferContext context = GetSingleTransferContext(checkpoint);
            CancellationTokenSource cancellationSource = new CancellationTokenSource();
            Console.WriteLine("\nTransfer started...\nPress 'c' to temporarily cancel your transfer...\n");
            Stopwatch stopWatch = Stopwatch.StartNew();
            Task task;
            ConsoleKeyInfo keyinfo;
            try
            {
                task = TransferManager.UploadAsync(localFilePath, blob, null, context, cancellationSource.Token);
                while (!task.IsCompleted)
                {
                    if (Console.KeyAvailable)
                    {
                        keyinfo = Console.ReadKey(true);
                        if (keyinfo.Key == ConsoleKey.C)
                        {
                            cancellationSource.Cancel();
                        }
                    }
                }
                await task;

// Metadata
               // IDictionary<string, string> blobdetails = new Dictionary<string, string>();
                //blobdetails.Add("department", "Opps");
                //blobdetails.Add("Canbedeleted", "False");
                //blobdetails.Add("Expirydate", DateTime.Now.ToShortDateString());
               // foreach (var key in blobdetails)
                //{
                  //  blob.Metadata[key.Key] = key.Value;
               // }
           }
            catch (Exception e)
            {
                Console.WriteLine("\nThe transfer is canceled: {0}", e.Message);
            }

            if (cancellationSource.IsCancellationRequested)
            {
                Console.WriteLine("\nTransfer will resume in 3 seconds...");
                Thread.Sleep(3000);
                checkpoint = context.LastCheckpoint;
                context = GetSingleTransferContext(checkpoint);
                Console.WriteLine("\nResuming transfer...\n");

                await TransferManager.UploadAsync(localFilePath, blob, null, context);
            }

            stopWatch.Stop();
            Console.WriteLine("\nTransfer operation completed in " + stopWatch.Elapsed.TotalSeconds + " seconds.");
            ExecuteChoice(account);
        }



Regards
Sunil Prabha

________________________________

**Comment by sunilprabha** [comment]
Hi , found the solution , I was missing few lines of code to make it works.  Thanks for the help and support. You may close the ticket.

 blob.FetchAttributes();
                IDictionary<string, string> blobdetails = new Dictionary<string, string>();

                blobdetails.Add("Canbedeleted", "False");

            blobdetails.Add("testdate", DateTime.Now.ToShortDateString());


                foreach (var key in blobdetails)
                {
                    blob.Metadata[key.Key] = key.Value;
                }
               await blob.SetMetadataAsync();


Regards
Sunil Prabha
404 542 5536
________________________________

**Comment by amnguye** [comment]
Hi, I just wanted to point out that you're using the v11 library and not the v12 library. I would suggest using `Azure.Storage.Blobs` because the library version you are currently using is deprecated. The original snippet of code you provided does not include the snippet I included at all.

As a last minute suggestion, I would still go with the first suggestion of setting the metadata when you're uploading the blob because it would do it in one request, instead of two separate requests.

Closing this issue as it seems to be resolved.
